keps/sig-node/4815-dra-partitionable-devices/README.md

```diff
@@ -0,0 +1,2069 @@
+            return nil, fmt.Errorf("duplicate mixin name detected: %s", m.Name)
+        }
+        mixinMap[m.Name] = &m
+    }
+
+    // Flatten each device and add it back to the flattened spec.
+    for _, d := range s.Devices {
+        flattened := &DeviceMixin{
+            Composite: &CompositeDeviceMixin{
+                Attributes: make(map[QualifiedName]DeviceAttribute),
+                Capacity:   make(map[QualifiedName]DeviceCapacity),
+            },
+        }
+        for _, m := range d.Composite.Includes {
+            for k, v := range mixinMap[m.Name].Composite.Attributes {
+                flattened.Composite.Attributes[k] = v
+            }
+            for k, v := range mixinMap[m.Name].Composite.Capacity {
+                flattened.Composite.Capacity[k] = v
+            }
+        }
+        for k, v := range d.Composite.Attributes {
+            flattened.Composite.Attributes[k] = v
+        }
+        for k, v := range d.Composite.Capacity {
+            flattened.Composite.Capacity[k] = v
+        }
+
+        d.Composite.Includes = nil
+        d.Composite.Attributes = flattened.Composite.Attributes
+        d.Composite.Capacity = flattened.Composite.Capacity
+
+        flattenedSpec.Devices = append(flattenedSpec.Devices, d)
+    }
+
+    return &flattenedSpec, nil
+}
+```
+
+The actual code will need to consider the fact that mixins are global to a pool
+across different `ResourceSlices`, but the general algorithm will be the same.
+
+### Defining device partitions in terms of consumed capacity in a composite device
+
+A simple example demonstrating how the `ConsumesCapacityFrom` field can be used
+to define multiple, allocatable partitions of a single overaching device can be
+seen below.
+
+```yaml
+devices:
+- name: gpu-0
+  composite:
+    capacity:
+      memory:
+        quantity: 40Gi
+- name: gpu-0-partition-0
+  composite:
+    capacity:
+      memory:
+        quantity: 10Gi
+    consumesCapacityFrom:
+    - name: gpu-0
+- name: gpu-0-partition-1
+  composite:
+    capacity:
+      memory:
+        quantity: 10Gi
+    consumesCapacityFrom:
+    - name: gpu-0
+- name: gpu-0-partition-2
+  composite:
+    capacity:
+      memory:
+        quantity: 10Gi
+    consumesCapacityFrom:
+    - name: gpu-0
+- name: gpu-0-partition-3
+  composite:
+    capacity:
+      memory:
+        quantity: 10Gi
+    consumesCapacityFrom:
+    - name: gpu-0
+```
+
+In this example, five devices are defined: a full GPU called "gpu-0" and four
+partitions of "gpu-0" called "gpu-0-partion-0", "gpu-0-partion-1",
+"gpu-0-partion-2", and "gpu-0-partion-3" respectively.
+
+In general, the way to interpret each of these device definitions is as
+follows:
+
+  * A device which *does not* have a `ConsumesCapacityFrom` field is assumed to
+    be a "source" of capacity from which other devices can pull when they are
+    being allocated.
+
+  * A device which *does* have a `ConsumesCapacityField` is assumed to be a
```

Comment:
What about a mixin that has capacities in it, will those get "consumedFrom"? I guess they will.

Also, it seems a device cannot have both "local" capacities and capacities from another "source"? I suspect that will constrain the utility of this. For example, I want to use this to represent disparate device types that are allocatable as a unit. I would effectively join multiple devices as a sort of "device molecule" and the resulting singular unit would advertise the union of all their attributes and capacities. I had planned to use a shared capacity pool to implement something like that. I am not sure if I can with this. I think maybe I can? See https://github.com/kubernetes-sigs/wg-device-management/pull/27#issuecomment-2155167940

Reply:
Patrick and I discussed this in a thread above that I can't seem to find anymore. One way to handle this would be to have a "meta" device that such a device pulls from that can't be allocated on its own. This would need to be assured by restricting it from selection in the device class.

My "original" proposal from many months ago decoupled local capacities from shared ones, but @thockin didn't like it because it was too abstract. In that proposal shared capacities were never allocatable on their own and always had to be consumed by some device. So a full GPU would need to pull from a shared capacity the same as its MIG devices would. That way they could all define local capacities (if they wanted) or just pull from the shared capacity.

However, it meant that there was some abstract notion of "shared capacity" not tied to any device, and @thockin wanted to make it clear that it was pulled from *some* device, not some abstract entity.

I'm happy to revisit this, but it's taken the shape it has based on this previous feedback.

Reply:
It's ok. I think I can make this work, and we can evolve it later if needed.